design-proposals/dra.md

```diff
@@ -10,23 +10,115 @@ control of their devices using Virtual Machines and Containers.
 - As a user, I would like to use KubeVirt's default driver
+- As a developer, I would like APIs to be extensible so I can develop drivers/webhooks/automation for custom use-cases
 
 ## Repos
 kubevirt/kubevirt
 
 # Design
 
-## API Examples
+At the high level there are two sets of changes needed for integrating DRA with KubeVirt:
+1. How DRA resources can be consumed by users of KubeVirt?
+2. What changes are needed for existing device-plugin drivers to be DRA compatible?
+
+This design focuses on part 1 of the problem. 
+
+For allowing users to consume DRA devices, there are two main changes needed:
+1. API changes
+2. Driver Implementation
+
+## API Changes
+
+```
+type VirtualMachineInstanceSpec struct {
+    // ResourceClaims defines which ResourceClaims must be allocated
+    // and reserved before the virt-launcher pod and hence the VMI is allowed to start. The resources
+    // will be made available to the domain which consume them
+    // by name.
+    //
+    // This is an alpha field and requires enabling the
+    // DynamicResourceAllocation feature gate in kubernetes
+    //  https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/
+    //
+    // This field is immutable.
+    //
+    // +featureGate=DynamicResourceAllocation
+    ResourceClaims []k8sv1.PodResourceClaim `json:"resourceClaims,omitempty"`
+}
+
+type GPU struct {
+	// Name of the GPU device as exposed by a device plugin
+	Name              string       `json:"name"`
+	DeviceName        string       `json:"deviceName"`
+	ClaimName         string       `json:"claimName"`
+	VirtualGPUOptions *VGPUOptions `json:"virtualGPUOptions,omitempty"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag   string `json:"tag,omitempty"`
+	// Claim is the name of resource, defined in spec.resourceClaims,
+	// that is used to provision this GPU device.
+	//
+	// This is an alpha field and requires enabling the
+	// DynamicResourceAllocation feature gate.
+	//
+	// This field is immutable. It can only be set for containers.
+	//
+	// +featureGate=DynamicResourceAllocation
+	// +optional
+	Claim string `json:"claim,omitempty"`
+}
+
+type HostDevice struct {
+	Name string `json:"name"`
+	// DeviceName is the resource name of the host device exposed by a device plugin
+	DeviceName string `json:"deviceName"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag string `json:"tag,omitempty"`
+	// Claim is the name of resource, defined in spec.resourceClaims,
+	// that is used to provision this host device.
+	//
+	// This is an alpha field and requires enabling the
+	// DynamicResourceAllocation feature gate.
+	//
+	// This field is immutable. It can only be set for containers.
+	//
+	// +featureGate=DynamicResourceAllocation
+	// +optional
+	Claim string `json:"claim,omitempty"`
+}
```

Comment:
syncing with Varun offline, we decided that a vmi.Status section for mapping the resource claim name to the name used in vmi spec will be needed. This mapping will be used by virt-launcher to generate the env variable